### PR TITLE
query-level caching for hive tables and iceberg table statistics

### PR DESCRIPTION
**Pull request by clemensvonschwerin**
We experienced long planning times due to expensive calls to hive metastore and s3 (during BaseTableScans). Caching of statistics and hive tables on the query-level reduced planning times by around 50% for us.

fixes https://github.com/trinodb/trino/issues/8675

### CODE REVIEW

**Review by findepi — CHANGES_REQUESTED** [review]
Last time we spoke, @electrum was very against query level caches, based on the hope they are not needed.

See the concern about second call to getTable here https://github.com/trinodb/trino/pull/8151#discussion_r643707727
and see test results for a test @homar just added here https://github.com/trinodb/trino/blob/30f2ecce1dd1fe787b075ecf21d30fe89d38e315/plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergMetastoreAccessOperations.java#L106-L108

The next step should be to understand _why_ we're doing those additional accesses to metastore.
Maybe we do not need them at all? Then, maybe we do not need the cache still?

cc @joshthoward

**Comment by electrum** [comment]
I am specifically against a generic metastore cache, like we have in the Hive connector. We can cache table statistics if we need it, though I'd like to understand why it is called multiple times, and if we can fix it in the engine. Requiring every connector to cache a likely expensive operation seems like a suboptimal design.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 1]
```diff
@@ -1330,4 +1363,26 @@ public long getSnapshotId()
...
     }
+
+    private static final class TableCacheLoader
+            extends CacheLoader<String, ConcurrentHashMap<SchemaTableName, Table>>
+    {
+        @Override
+        public ConcurrentHashMap<SchemaTableName, Table> load(final String unused)
+                throws Exception
+        {
+            return new ConcurrentHashMap<>();
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:1374]: I think i don't understand what this is doing.

Anyway, i guess this will go away once we move away from unnecessary `Cache` use.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 2]
```diff
@@ -221,6 +228,14 @@ public IcebergMetadata(
         this.commitTaskCodec = requireNonNull(commitTaskCodec, "commitTaskCodec is null");
         this.tableOperationsProvider = requireNonNull(tableOperationsProvider, "tableOperationsProvider is null");
         this.trinoVersion = requireNonNull(trinoVersion, "trinoVersion is null");
+        this.tableCache = CacheBuilder.newBuilder()
+                .maximumSize(100)
+                .expireAfterWrite(5, TimeUnit.MINUTES)
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:233]: The IcebergMetadata object is per-query scoped, so you don't need time-based eviction here.
That's why `tableMetadataCache` uses ordinary map.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 3]
```diff
@@ -221,6 +228,14 @@ public IcebergMetadata(
...
+                .build(new TableCacheLoader());
+        this.tableStatisticsCache = CacheBuilder.newBuilder()
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:235]: if adding stats cache is an improvement, i would expect this to be reflected by a test change.
Currently i see `TestIcebergMetadataFileOperations` passes without modification.

Let's move statistics to separate PR (or at least separate commit), so that we can focus on metastore interactions here.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 4]
```diff
@@ -293,6 +308,19 @@ public IcebergTableHandle getTableHandle(ConnectorSession session, SchemaTableNa
                 TupleDomain.all());
     }
 
+    Optional<Table> getHiveTable(ConnectorSession session, SchemaTableName schemaTableName)
+    {
+        var queryTableCache = tableCache.getUnchecked(session.getQueryId());
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:313]: see https://github.com/trinodb/trino/blob/master/DEVELOPMENT.md#avoid-var

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 5]
```diff
@@ -201,6 +205,9 @@
 
     private final Map<String, Optional<Long>> snapshotIds = new ConcurrentHashMap<>();
     private final Map<SchemaTableName, TableMetadata> tableMetadataCache = new ConcurrentHashMap<>();
+
+    private final LoadingCache<String, ConcurrentHashMap<SchemaTableName, Table>> tableCache;
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:209]: Nested keying `query id -> SchemaTableName -> Table` is redundant.
Make `SchemaTableName` the key here, as in `tableMetadataCache` (and effectively in `snapshotIds`, although the code doesn't make it obvious)

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java (hunk 1 above)
Comment (clemensvonschwerin) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:1374]: The cache builder needs a function as an argument which creates the cache entry if there is a miss for a certain key. Here we simply create a new ConcurrentHashMap if there is no cache for a certain query id, yet.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java (hunk 2 above)
Comment (clemensvonschwerin) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:233]: Ah, I did not know that. Then we do not even need the two levels query -> table but one level is enough.

**Comment by colebow** [comment]
:wave: @clemensvonschwerin - this PR is inactive and doesn't seem to be under development. If you'd like to continue work on this at any point in the future, feel free to re-open.

**Comment by bitsondatadev** [comment]
> I like the engine-side approach: #12196 Engine shouldn't ask a connector for same information multiple times.

This was completed here: https://github.com/trinodb/trino/pull/13047
